### ISSUE TITLE
Rendering partial over collection with `cached: true` leave out same items

### ISSUE THREAD

**Issue by foton** [rails/rails#35114]
### Steps to reproduce
I have view with:
```
<% same_sets = [
  [1,2,3,4,5],
  [1,2,3,4,5],
  [1,2,3,4,4],
  [1,2,3,4,4],
  [1,2,3,4,6],
]%>
with 'cached: true'<br />
<%= render partial: 'test_row', collection: same_sets, cached: true %>
<br />
with 'cached: false'<br />
<%= render partial: 'test_row', collection: same_sets, cached: false %>
<br />
```
and corresponding partial `_test_row.html.erb` : 
```
<%= test_row.first %> | <%= test_row.second %> | <%= test_row.third %> | <%= test_row.fourth %> | <%= test_row.fifth %><br />
```

That view will produce this result: 
```
with 'cached: true'
1 | 2 | 3 | 4 | 5
1 | 2 | 3 | 4 | 4
1 | 2 | 3 | 4 | 6

with 'cached: false'
1 | 2 | 3 | 4 | 5
1 | 2 | 3 | 4 | 5
1 | 2 | 3 | 4 | 4
1 | 2 | 3 | 4 | 4
1 | 2 | 3 | 4 | 6
```

### Expected behavior

For array with size 5, I expect 5 rows, with or without caching.
### Actual behavior

For array with size 5, I got 3 (different) rows with caching enabled.

### System configuration
**Rails version**: 5.2.2

**Ruby version**:
2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]

**Comment by st0012** [comment]
I think this is a bug and I can reproduce it in latest master as well.

**Comment by y-yagi** [comment]
Fixed by #35145
